### DISCUSSION TITLE
DIS output

### DISCUSSION THREAD

**Discussion by UcfAerokid** [JSBSim-Team/jsbsim#576] [Show and tell]
I've implemented the KDIS Open Source library to add DIS (IEEE Std 1278.1A-1998) output from JSBSim for visualization and enable participation in distributed simulation events.  It's a very basic implementation of position and orientation, with body frame-referenced dead reckoning to enable smooth representation while keeping DIS traffic to a minimum.

The XML output element is shown below:
`<output type="DIS" name="127.0.0.1" port="61211" entityId="1:3:5" rate="5" deadReckoning="on" exerciseId="3" entityType="1:2:225:1:2:3:0">	</output>`

The entity type is derived from the standard's enumerations document (SISO-REF-010-XXXX, Enumeration and Bit Encoded Values for Use with Protocols for Distributed Interactive Simulation Applications).  The Entity ID and exercise ID are necessary to identify the simulation entity within the applicable simulation exercise event.

See attached .zip file for relevant source of derived output type class and a Wireshark PCAP of the resulting output a the specified 5Hz rate with dead reckoning implemented.

Of course, the src/models/FgOutput.cpp has to have the type implemented in FGOutput::Load()
`else if (type == "DIS") {
	  Output = new FGOutputDisSocket(FDMExec);` 

I have some TODO's in the source .cpp file.  The most important would be an array of <articulation> elements to map DIS articulation indices to JSBSim properties so that model can show control surface deflections, etc.  Also appearance bits to reflect nav/strobe lights on/off status, etc.

My JSBSim VS Solution links to my KDIS tree, which can link in static or DLL lib, but I'm not sure how the maintainers would approach implementing other open sources.

Welcome any feedback and questions.

[jsbsimDisOut.zip](https://github.com/JSBSim-Team/jsbsim/files/7915743/jsbsimDisOut.zip)

**Comment by seanmcleod** [comment]
> I've implemented the KDIS Open Source library to natively enable DIS

Don't quite follow what you're trying to say. Are you saying you've looked at the DIS standard, IEEE 1278 and written your own library to implement at least the portion that deals without outputting position and orientation?

Or do you mean you've 'implemented' some other existing DIS open source library - https://sourceforge.net/p/kdis/wiki/Main_Page/ by compiling it locally and linking it in?

Or maybe both and @UcfAerokid == Karl Jones (https://sourceforge.net/u/karljj1/profile/)?

**Reply by UcfAerokid** [reply]
Thanks for the question.  I'm not Karl, but I implemented his library to add DIS output to JSBSim, similar to the sockets and FG output.  KDIS is the most mature and complete DIS library I've used and implemented over the years and it enabled me to add DIS output with minimal work and just those few additions/modifications to the JSBSim code.

His source builds both static and dynamic libraries and I linked it in dynamically.

**Comment by UcfAerokid** [comment]
Another observation is that it also outputs at every trim step, which is undesired behavior.  Will have to find the right triggers to skip the output step if the sim is trimming.

**Comment by seanmcleod** [comment]
> to skip the output step if the sim is trimming.

https://github.com/JSBSim-Team/jsbsim/blob/d53cb46461c2e47fb87a44b7273e1a8fc86bffdc/src/models/flight_control/FGActuator.cpp#L144-L148

**Reply by UcfAerokid** [reply]
Thank you!  I assume TRUE means it's trimming.
